### PR TITLE
Extract DP3T and inject it

### PR DESCRIPTION
**Pull request by dcampogiani**
The goal of this PR is to extract the DP3T interface to reduce coupling with the SDK and improve the architecture.

This way we can start to think about testing and we can avoid passing Application/Context in several places.

**Dependency Injection**
I've used Dagger to inject dependencies, if you are not an android dev, here is what you need to understand this PR:
 - AppComponent is the entry point of injection
 - A module is a way to provide dependencies
 - AppComponent is resolving all dependencies using the module you provide (AppModule and ViewModelBuilder in this case)
 - On Android Activities and Fragment we can't use constructor injection so in the appropriate callback (onCreate/onAttach) we get a reference to AppComponent (hosted by MainApplication) and ask for member injection
 - On ViewModels we can use constructor injection, but we need to tell the Framework how to resolve dependencies. For this purpose, ViewModelFactory is glue code between Dagger and Android ViewModels

**Coroutines**
I've used Kotlin coroutines (both suspended method and Flows) because of [Structured concurrency](https://medium.com/@elizarov/structured-concurrency-722d765aa952). Basically it will allow us to use async code with an automatic unsubscription, avoiding leaks.

If you have any question please ask.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -5,70 +5,88 @@
...
+    implementation 'androidx.lifecycle:lifecycle-runtime-ktx:2.2.0'
+    implementation 'androidx.viewpager2:viewpager2:1.0.0'
 
-	implementation 'androidx.appcompat:appcompat:1.1.0'
-	implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
-	implementation 'androidx.fragment:fragment:1.2.4'
-	implementation 'androidx.lifecycle:lifecycle-viewmodel:2.2.0'
-	implementation 'androidx.lifecycle:lifecycle-livedata:2.2.0'
-	implementation 'androidx.viewpager2:viewpager2:1.0.0'
+    implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.5'
```
Comment (Harrypulvirenti) [inline app/build.gradle:73]: You don't need `coroutines-android`?

File: app/build.gradle [hunk 2]
```diff
@@ -5,70 +5,88 @@
...
 
-	implementation 'io.reactivex.rxjava3:rxandroid:3.0.0'
-	implementation 'io.reactivex.rxjava3:rxjava:3.0.0'
+    implementation "com.google.dagger:dagger:2.27"
+    kapt "com.google.dagger:dagger-compiler:2.27"
```
Comment (Harrypulvirenti) [inline app/build.gradle:76]: You can define a version for groups of dependencies like in this case we can create a `daggerVersion` var

File: app/src/main/java/org/dpppt/android/app/contacts/ContactsFragment.java [hunk 3]
```diff
@@ -16,113 +17,124 @@
...
+    }
+
+    public ContactsFragment() {
+        super(R.layout.fragment_contacts);
+    }
+
+    @Override
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
+        MainApplication.Companion.getAppComponent(context).inject(this);
```
Comment (Harrypulvirenti) [inline app/src/main/java/org/dpppt/android/app/contacts/ContactsFragment.java:60]: `BaseInjectFragment`, extensions or one simple util to do this?

File: app/src/main/java/org/dpppt/android/app/contacts/ContactsFragment.java [hunk 4]
```diff
@@ -16,113 +17,124 @@
...
+    }
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        tracingViewModel = new ViewModelProvider(requireActivity(), viewModelFactory).get(TracingViewModel.class);
```
Comment (Harrypulvirenti) [inline app/src/main/java/org/dpppt/android/app/contacts/ContactsFragment.java:66]: same here try to wrap as much as you can inside utils methods in order to ensure that everyone can benefit your implementation.

**Comment by dcampogiani** [comment]
I would like to close this PR.

I don't like how a member of this project is attacking others on LinkedIn, and I don't want to be part of it.

Thanks for your time and your effort.

**Pull request closed without merging**